include/seqan3/alignment/data_structure/aligned_sequence_concept.hpp

```diff
@@ -0,0 +1,258 @@
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \brief Includes the aligned_sequence_concept and the related insert_gap and
+ *        erase_gap functions to enable stl container support.
+ * \author Svenja Mehringer <svenja.mehringer AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <seqan3/alignment/exception.hpp>
+#include <seqan3/alphabet/gap/all.hpp>
+#include <seqan3/std/concept/container.hpp> // sequence_container_concept
+#include <seqan3/std/concept/range.hpp>     // forward_range_concept
+#include <seqan3/core/metafunction/all.hpp> // value_type_t
+
+namespace seqan3
+{
+
+/*!\interface seqan3::aligned_sequence_concept <>
+ * \extends   seqan3::forward_range_concept
+ * \brief     The generic concept for an aligned sequence.
+ * \ingroup   data_structure
+ *
+ * \details
+ *
+ * This concept describes the requirements a sequence must fulfill
+ * in order to be part of the seqan3::alignment object.
+ */
+/*!\name Requirements for seqan3::aligned_sequence_concept
+ * \brief You can expect these functions on all types that model the seqan3::aligned_sequence_concept.
+ *
+ * \details
+ *
+ * Type Requirements:
+ *   1. The type must module the seqan3::forward_range_concept.
+ *   2. The `value_type` must module the alphabet concept.
+ *   3. The `value_type` must be assignable from seqan3::gap::GAP.
+ *
+ * Furthermore, the free functions seqan3::aligned_sequence_concept::insert_gap
+ * and seqan3::aligned_sequence_concept::erase_gap exist for the type.
+ * \{
+ */
+/*!\fn      inline void insert_gap(seq_type & seq, typename seq_type::const_iterator pos_it)
+ * \brief   Insert a seqan3::gap::GAP into a sequence container.
+ * \relates seqan3::aligned_sequence_concept
+ * \tparam        seq_type The sequence container type. Must model the
```

Comment:
Yes for the concept, but for the generic `insert_gap` function I require the sequence container concept because I need the function `insert` to modify the sequence.

Reply:
Yes, but this is the concept definition and not the free function!